File: c/tree/bst.c
#include <stdio.h>
#include <stdlib.h>
struct node {
     int data;
     struct node* left;
     struct node* right;
};
struct node* newNode(int data) {
     struct node* node = (struct node*) malloc(sizeof(struct node));
     node->data = data;
     node->left = NULL;
     node->right = NULL;
     return node;
}
void push(struct node ** root, int data) {
    if (*root == NULL) {
        struct node * child = newNode(data);
        *root = child;
        return;
    }
    if ((*root)->data > data) {
        push(&(*root)->left, data);
    } else {
        push(&(*root)->right, data);
    }
}
void printInOrder(struct node* root) {
     if (root == NULL)
          return;
     printInOrder(root->left);
     printf("%d ", root->data);  
     printInOrder(root->right);
}
void printPreOrder(struct node* root) {
     if (root == NULL)
          return;
     printf("%d ", root->data);
     printPreOrder(root->left);
     printPreOrder(root->right);
}
void printPostOrder(struct node* root) {
     if (root == NULL)
          return;
     printPostOrder(root->left);
     printPostOrder(root->right);
     printf("%d ", root->data);
}
int main() {
    struct node *root  = NULL;
    push(&root, 2);
    push(&root, 3);
    push(&root, 4);
    push(&root, 5);
    push(&root, 1);
    printf("InOrder traversal of binary tree is \n");
    printInOrder(root);
    printf("\n");
    printf("PreOrder traversal of binary tree is \n");
    printPreOrder(root);
    printf("\n");
    printf("PostOrder traversal of binary tree is \n");
    printPostOrder(root);
    printf("\n");
    return 0;
}

File: static/js/bst.js
(function() {
//bst: Binary Search Tree
var Node = (function(){
    function Node(nodeVal) {
        this.data = nodeVal;
        this.left = null;
        this.right = null;
    }
    return Node;
})();
var BinarySearchTree = (function() {
    function BinarySearchTree() {
        this.root = null;
    }
    BinarySearchTree.prototype.insertNode = function(node, newNode) { 
        if(newNode.data < node.data) {
            if(node.left === null) {
                node.left = newNode; 
            } else {
                this.insertNode(node.left, newNode);  
            } 
        } else {
            if(node.right === null) {
                node.right = newNode; 
            } else {
                this.insertNode(node.right, newNode); 
            }
        }
    };
    BinarySearchTree.prototype.printInOrder = function(node) {
        if (node != null) {
            this.printInOrder(node.left);
            console.log(node.data);
            this.printInOrder(node.right);
        }
    };
    BinarySearchTree.prototype.printPostOrder = function(node) {
        if (node != null) {
            this.printInOrder(node.left);
            this.printInOrder(node.right);
            console.log(node.data);
        }
    };
    BinarySearchTree.prototype.printPreOrder = function(node) {
        if (node != null) {
            console.log(node.data);
            this.printInOrder(node.left);
            this.printInOrder(node.right);
        }
    }
    return BinarySearchTree;
})();
var bst = new BinarySearchTree();
var BST = function(selector, context) {
    return new BST.fn.init(selector, context);
};
BST.fn = BST.prototype = {
    constructor: BST,
    init: function(selector, context) {
        return this;
    }
};
BST.fn.init.prototype = BST.fn;
BST.extend = BST.fn.extend = function(options) {
    for (var key in options) {
        this[key] = options[key];
    }
    return this;
};
BST.extend({
    insert: function(data) {
        var newNode = new Node(data);
        if (bst.root == null) {
            bst.root = newNode;
        } else {
            bst.insertNode(bst.root, newNode);
        }
        return data;
    },
    printInOrder: function() {
        bst.printInOrder(bst.root);
        return 0;
    },
    printPreOrder: function() {
        bst.printPreOrder(bst.root);
        return 0;
    },
    printPostOrder: function() {
        bst.printPostOrder(bst.root);
        return 0;
    }
});
window.BST = window.$BST = BST;
})();
